keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -557,278 +532,231 @@ reconstructed by the driver. This has no effect on already allocated claims
 
-```yaml
-kind: ResourceSlice
-apiVersion: resource.k8s.io/v1alpha2
-...
-spec:
-  nodeName: worker-1
-  driverName: cards.dra.example.com
-  namedResources:
-    ...
-```
-
-Such a model could be created to represent resources in a manner similar to the
-opaque strings passed over the tradition device plugin API to the kubelet. The
-one addition being that each named resource can have a set of arbitrary
-attributes attached to it.
-
-If a driver wanted to use a different structured model to represent its resources,
-a new structured model would need to be defined inside Kuberenetes, and a field
-would need to be added to this struct at the same level as
-`namedResources`. Driver implementors would then have the option
-to set this new field instead.
-
-**Note:** If a new model is added to the schema but clients are not updated,
-they'll encounter an object with no information from any known structured model
-when they serialize into their known version of a `ResourceSlice`. This
-tells them that they cannot handle the object because the API has been extended.
-
-Drivers can use different structured models by publishing multiple
-`ResourceSlice` objects, as long as each model represents a distinct set of
-resources. Whether the information about resources of one particular structured
-model must fit into one ResourceSlice object (or be distributed across
-many) depends on how that particular structured model describes its resources. In
-all cases, the size of each object is a hard limit and one must take this into
-account when designing a structured model and preparing ResourceSlice objects
-for it.
-
-Below is an example of a driver that provides two discrete GPU cards using the
-`namedResources` model described above:
+Embedded inside each `ResourceSlice` is a list of one or more devices, represented as a named list of attributes.
 
 ```yaml
 kind: ResourceSlice
-apiVersion: resource.k8s.io/v1alpha2
+apiVersion: resource.k8s.io/v1alpha3
 ...
 spec:
+  # The node name indicates the node.
+  # Each driver on a node provides pools of devices for allocation,
+  # with unique device names inside each pool.
+  # Usually, but not necessarily, that pool name is the same as the
+  # node name.
   nodeName: worker-1
+  poolName: worker-1
   driverName: cards.dra.example.com
-  namedResources:
-  - name: gpu-0
+  devices:
+  - name: card-1 # unique inside the worker-1 pool
     attributes:
-    - name: UUID
-      string: GPU-ceea231c-4257-7af7-6726-efcb8fc2ace9
-    - name: driverVersion
-      string: 1.2.3
-    - name: runtimeVersion
-      string: 11.1.42
-    - name: memory
-      quantity: 16Gi
+    - name: manufacturer # a vendor-specific attribute, automatically qualified as cards.dra.example.com/manufacturer
+      string: Cards-are-us Inc.
     - name: productName:
-      string: ACME T1000 32GB
-  - name: gpu-1
-    attributes:
-    - name: UUID
-      string: GPU-6aa0af9e-a2be-88c8-d2b3-2240d25318d7
+      string: ACME T1000 16GB
```

Comment:
We intentionally make them *not* include the `Value` suffix. This is because we would have preferred not to have the `Value` suffix on the go type at all. There was a long discussion on why we had to do this at one point, which I'm sure Patrick has a reference to somewhere.

Reply:
We wanted the YAML names to be short, but then had to add the `Value` suffix in the Go code because API types must have a `String()` method and thus cannot have a `String` field.

There's a comment about that above the Go fields.